ocaml/testsuite/tests/typing-unique/overwriting.ml

```diff
@@ -661,6 +787,169 @@ Error: Overwrite may not change the tag to OptionA.
 |}]
 
+type 'a mutable_record = { mutable m : 'a }
+
+let mutable_field_aliased r =
+  let y = OptionA "foo" in
+  r.m <- y;
+  (unique_ r), y
+[%%expect{|
+type 'a mutable_record = { mutable m : 'a; }
+val mutable_field_aliased :
+  unique_ options mutable_record -> options mutable_record * options @@
+  global many = <fun>
+|}]
+
+let mutable_field_aliased r =
+  unique_ r.m
+[%%expect{|
+val mutable_field_aliased : unique_ 'a mutable_record -> 'a @@ global many =
+  <fun>
+|}]
+
+let tag_of_mutable_field r =
+  match r with
+  | { m = OptionA s } ->
+    overwrite_ r.m with OptionA s
+  | _ -> OptionB ""
+[%%expect{|
+Line 4, characters 4-33:
+4 |     overwrite_ r.m with OptionA s
+        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Alert Translcore: Overwrite not implemented.
+Uncaught exception: File "ocaml/parsing/location.ml", line 1106, characters 2-8: Assertion failed
+
+|}]
+
+let mutating_tag_seq r =
+  match r with
+  | { m = OptionA s } ->
+    r.m <- OptionB s;
+    overwrite_ r.m with OptionA s
+  | _ -> OptionB ""
+[%%expect{|
+Line 5, characters 24-31:
+5 |     overwrite_ r.m with OptionA s
+                            ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_seq_parent r =
+  match r with
+  | { m = { x = OptionA s } } ->
+    r.m <- { x = OptionB s };
+    overwrite_ r.m.x with OptionA s
+  | _ -> OptionB ""
+[%%expect{|
+Line 5, characters 26-33:
+5 |     overwrite_ r.m.x with OptionA s
+                              ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_par r =
+  match r with
+  | { m = OptionA s } ->
+    (r.m <- OptionB s), overwrite_ r.m with OptionA s
+  | _ -> (), OptionB ""
+[%%expect{|
+Line 4, characters 44-51:
+4 |     (r.m <- OptionB s), overwrite_ r.m with OptionA s
+                                                ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_par_parent r =
+  match r with
+  | { m = { x = OptionA s } } ->
+    (r.m <- { x = OptionB s }), overwrite_ r.m.x with OptionA s
+  | _ -> (), OptionB ""
+[%%expect{|
+Line 4, characters 54-61:
+4 |     (r.m <- { x = OptionB s }), overwrite_ r.m.x with OptionA s
+                                                          ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_choice r =
+  match r with
+  | { m = OptionA s } ->
+    if true then (r.m <- OptionB s; r.m)
+            else overwrite_ r.m with OptionA s
+  | _ -> OptionB ""
+[%%expect{|
+Line 5, characters 37-44:
+5 |             else overwrite_ r.m with OptionA s
+                                         ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
```

Comment:
This is a bit sad. The mutation definitely didn't happen: it's in the opposite branch.